ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/decoders/TomlAbstractDecoder.kt

```diff
@@ -49,25 +51,62 @@ public abstract class TomlAbstractDecoder : AbstractDecoder() {
 
     internal abstract fun decodeKeyValue(): TomlKeyValue
 
-    private fun invalidType(typeName: String, requiredType: String): Nothing {
-        val keyValue = decodeKeyValue()
-        throw IllegalTypeException(
-            "<$typeName> type is not allowed by toml specification," +
-                    " use <$requiredType> instead" +
-                    " (key = ${keyValue.key.content}; value = ${keyValue.value.content})", keyValue.lineNo
-        )
-    }
-
+    /**
+     * This is just an adapter from `kotlinx.serialization` to match the content with a type from a Toml Tree,
+     * that we have parsed to a type that is described in user's code. For example:
+     * >>> input: a = "5"
+     * >>> stored in Toml Tree: TomlString("5")
+     * >>> expected by user: data class A(val a: Int)
+     * >>> TomlString cannot be cast to Int, user made a mistake -> IllegalTypeException
+     */
     private inline fun <reified T> decodePrimitiveType(): T {
         val keyValue = decodeKeyValue()
         try {
-            return keyValue.value.content as T
+            return when (val value = keyValue.value) {
+                is TomlLong -> decodeInteger(value.content as Long, keyValue.lineNo)
+                else -> keyValue.value.content as T
+            }
         } catch (e: ClassCastException) {
-            throw CastException(
+            throw IllegalTypeException(
                 "Cannot decode the key [${keyValue.key.content}] with the value [${keyValue.value.content}]" +
                         " with the provided type [${T::class}]. Please check the type in your Serializable class or it's nullability",
                 keyValue.lineNo
             )
         }
     }
+
+    /**
+     * After a lot of discussions, we have finally decided to allow to use Integer types and not only Long.
+     * This method does simple validation of integer values to avoid overflow. For example, you really want to use byte,
+     * we will check here, that your byte value does not exceed 127 and so on.
+     */
+    private inline fun <reified T> decodeInteger(content: Long, lineNo: Int): T = when (T::class) {
+        Byte::class -> validateAndConvertInt(content, lineNo, BYTE) { num: Long -> num.toByte() as T }
+        Short::class -> validateAndConvertInt(content, lineNo, SHORT) { num: Long -> num.toShort() as T }
+        Int::class -> validateAndConvertInt(content, lineNo, INT) { num: Long -> num.toInt() as T }
+        Long::class -> validateAndConvertInt(content, lineNo, LONG) { num: Long -> num as T }
+        else -> invalidType(T::class.toString(), "Signed Type")
+    }
+
+    private inline fun <reified T> validateAndConvertInt(
+        content: Long,
+        lineNo: Int,
+        limits: IntegerLimitsEnum,
+        conversion: (Long) -> T,
+    ): T = if (content in limits.min..limits.max) {
+        conversion(content)
+    } else {
```

Comment:
I think this code is based on my PR which used the MIN/MAX boundaries, but I checked the Kotlinx Serialization JSON decoder and I think it has a better way.

https://github.com/Kotlin/kotlinx.serialization/blob/d7e58c2eeb02e1da87ad36d41ef25523fafb6935/formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt#L299-L304

1. decode to Long
2. convert to `value.toInt()`
3. if the result is different, then throw an exception

It's less code, and I think easier to understand

Reply:
The only problem I see here is that comparing operation with limits can be easier than making a `toInt()` and than comparing... But anyway probably your variant is good enough.

Reply:
speaking about the idea - I actually asked a question on stack overflow about that and as well as I and you thought - everyone suggests to check boundaries: https://stackoverflow.com/questions/74510016/how-to-understand-if-there-was-an-integer-overflow-during-parsing-of-string-with :D 

So probably, it is a good idea to have the most obvious solution

Reply:
Personally I prefer the KxS JSON way with `.toInt()`, but the result is the same pick whatever you prefer :)

Reply:
Also funny thing that @NightEule5 suggested same solution as you here: https://github.com/akuleshov7/ktoml/pull/153#discussion_r1003881958 :D